### PR TITLE
Use OpenSSL 3's HKDF if it is available 

### PR DESCRIPTION
**Pull request by vcsjones**
If OpenSSL 3 is not available, or HKDF is not present, then the managed implementation will continue to be used.

### CODE REVIEW

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HKDF.cs [hunk 1]
```diff
@@ -63,19 +66,15 @@ public static int Extract(HashAlgorithmName hashAlgorithmName, ReadOnlySpan<byte
...
-            Debug.Assert(written == prk.Length, $"Bytes written is {written} bytes which does not match output length ({prk.Length} bytes)");
-        }
-
         /// <summary>
         /// Performs the HKDF-Expand function
         /// See section 2.3 of <a href="https://tools.ietf.org/html/rfc5869#section-2.3">RFC5869</a>
         /// </summary>
         /// <param name="hashAlgorithmName">The hash algorithm used for HMAC operations.</param>
-        /// <param name="prk">The pseudorandom key of at least <see cref="HashLength"/> bytes (usually the output from Expand step).</param>
+        /// <param name="prk">
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HKDF.cs:74]: These doc comments referred to a private method which got moved. I updated them to the documentation that is on docs.

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HKDF.cs [hunk 2]
```diff
@@ -84,10 +83,14 @@ private static void Extract(HashAlgorithmName hashAlgorithmName, int hashLength,
         public static byte[] Expand(HashAlgorithmName hashAlgorithmName, byte[] prk, int outputLength, byte[]? info = null)
         {
             ArgumentNullException.ThrowIfNull(prk);
-
             ArgumentOutOfRangeException.ThrowIfNegativeOrZero(outputLength);
 
-            int hashLength = HashLength(hashAlgorithmName);
+            int hashLength = Helpers.HashLength(hashAlgorithmName);
+
+            if (prk.Length < hashLength)
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HKDF.cs:90]: This validation was in a "core" method that got abstracted, so it got pushed up in to the public API surface.

File: src/native/libs/System.Security.Cryptography.Native/pal_evp_kdf.c [hunk 3]
```diff
@@ -157,3 +157,166 @@ int32_t CryptoNative_KbkdfHmacOneShot(
...
+    char* algorithm,
+    uint8_t* info,
+    int32_t infoLength,
+    uint8_t* destination,
+    int32_t destinationLength)
+{
+    return HkdfCore(
+        kdf,
+        "EXPAND_ONLY",
+        prk,
```
Comment (bartonjs) [inline src/native/libs/System.Security.Cryptography.Native/pal_evp_kdf.c:289]: I like the containment of the string literals here.

I do wonder if we should be static_asserting that they're equal to a named constant, or doing the ifdef / assert / else / define pattern so we can use it here symbolically

**Review by bartonjs — APPROVED** [review]

**Comment by vcsjones** [comment]
NO-MERGE while this is validated against SCOSSL.

File: src/native/libs/System.Security.Cryptography.Native/pal_evp_kdf.c (hunk 3 above)
Comment (vcsjones) [reply src/native/libs/System.Security.Cryptography.Native/pal_evp_kdf.c]: The strings as far as I can tell have no named constants. They are documented magic strings.

It does have constants that are integers. So I switched to that and did all the define assert stuff.

**Review by bartonjs — APPROVED** [review]

**Comment by bartonjs** [comment]
Note to future me: cryptograhic-docs-impact was removed because it's being incorporated in the 9.0 docs.

**Pull request merged**